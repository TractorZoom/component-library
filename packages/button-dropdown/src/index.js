module.exports = require('./button-dropdown');
