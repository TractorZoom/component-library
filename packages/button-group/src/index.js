module.exports = require('./button-group');
