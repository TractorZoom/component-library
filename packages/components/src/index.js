module.exports = {
    BasicDialog: require('@tractorzoom/basic-dialog'),
    Button: require('@tractorzoom/button'),
    ButtonDropdown: require('@tractorzoom/button-dropdown'),
    ButtonGroup: require('@tractorzoom/button-group'),
    Chip: require('@tractorzoom/chip'),
    EquipmentCard: require('@tractorzoom/equipment-card'),
    List: require('@tractorzoom/list'),
    NativeSelect: require('@tractorzoom/native-select'),
    NumberField: require('@tractorzoom/number-field'),
    Lozenge: require('@tractorzoom/lozenge'),
    PlacesSelect: require('@tractorzoom/places-select'),
    Radio: require('@tractorzoom/radio'),
    Select: require('@tractorzoom/select'),
    SidePanel: require('@tractorzoom/side-panel'),
    SimpleSelect: require('@tractorzoom/simple-select'),
    TextField: require('@tractorzoom/text-field'),
    Theme: require('@tractorzoom/theme'),
    UsageMeter: require('@tractorzoom/usage-meter'),
};
