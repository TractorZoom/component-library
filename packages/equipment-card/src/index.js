module.exports = require('./equipment-card');
