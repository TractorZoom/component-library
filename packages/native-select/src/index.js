module.exports = require('./native-select');
