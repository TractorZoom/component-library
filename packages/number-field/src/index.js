module.exports = require('./number-field');
