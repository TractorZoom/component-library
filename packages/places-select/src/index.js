const { PlacesSelect } = require('./places-select');
const utilities = require('./utilities');

module.exports = {
    PlacesSelect: PlacesSelect,
    getFullAddressString: utilities.getFullAddressString,
    getLatLon: utilities.getLatLon,
};
