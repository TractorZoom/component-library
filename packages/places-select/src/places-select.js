import usePlacesAutocomplete, { getGeocode, getLatLng } from 'use-places-autocomplete';
import Autocomplete from '@material-ui/lab/Autocomplete';
import Grid from '@material-ui/core/Grid';
import LocationOnIcon from '@material-ui/icons/LocationOn';
import PropTypes from 'prop-types';
import React from 'react';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import parse from 'autosuggest-highlight/parse';
import useStyles from './styles';

const PlacesSelect = (props) => {
    const classes = useStyles();
    const { address, setAddress } = props;

    const {
        clearSuggestions,
        suggestions: { status, data },
        setValue,
    } = usePlacesAutocomplete({ debounce: 300 });

    const handleInput = (e, newInputValue) => {
        setValue(newInputValue);
    };

    const handleSelect = async (e, val) => {
        setValue(val, false);

        let address = {};

        if (val) {
            address = { fullAddress: val.description };

            const geocodedAddress = await getGeocode({ address: val.description });
            const { lat, lng } = await getLatLng(geocodedAddress[0]);

            geocodedAddress[0].address_components.forEach((component) => {
                if (component.types.includes('street_number')) {
                    address.streetNumber = component.long_name;
                }

                if (component.types.includes('route')) {
                    address.street = component.short_name;
                }

                if (component.types.includes('locality')) {
                    address.city = component.short_name;
                }

                if (component.types.includes('administrative_area_level_1')) {
                    address.state = component.short_name;
                }

                if (component.types.includes('country')) {
                    address.country = component.short_name;
                }

                if (component.types.includes('postal_code')) {
                    address.zip = component.short_name;
                }
            });

            address = { ...address, address: `${address.streetNumber} ${address.street}`, lat, lng };
        }

        setAddress(address);
        clearSuggestions();
    };

    return (
        <Autocomplete
            {...props}
            getOptionLabel={(option) => (typeof option === 'string' ? option : option.description)}
            filterOptions={(x) => x}
            options={status === 'OK' ? [...data, { addPoweredByGoogle: true }] : []}
            autoComplete
            includeInputInList
            filterSelectedOptions
            value={address?.fullAddress}
            onChange={handleSelect}
            onInputChange={handleInput}
            renderInput={(params) => <TextField {...params} label='Location' variant='outlined' fullWidth />}
            renderOption={(option) => {
                if (option.addPoweredByGoogle) {
                    return (
                        <Grid key='poweredByGoogle' container justify='flex-end'>
                            <Grid item>
                                <img
                                    src='https://developers.google.com/maps/documentation/images/powered_by_google_on_white.png'
                                    alt='Powered by Google'
                                />
                            </Grid>
                        </Grid>
                    );
                }

                const matches = option.structured_formatting.main_text_matched_substrings;
                const parts = parse(
                    option.structured_formatting.main_text,
                    matches.map((match) => [match.offset, match.offset + match.length])
                );

                return (
                    <Grid container alignItems='center'>
                        <Grid item>
                            <LocationOnIcon className={classes.icon} />
                        </Grid>
                        <Grid item xs>
                            {parts.map((part, index) => (
                                <span key={index} style={{ fontWeight: part.highlight ? 700 : 400 }}>
                                    {part.text}
                                </span>
                            ))}

                            <Typography variant='body2' color='textSecondary'>
                                {option.structured_formatting.secondary_text}
                            </Typography>
                        </Grid>
                    </Grid>
                );
            }}
        />
    );
};

PlacesSelect.propTypes = {
    address: PropTypes.string.isRequired,
    setAddress: PropTypes.func.isRequired,
};

export default PlacesSelect;
