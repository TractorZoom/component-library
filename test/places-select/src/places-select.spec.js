import { PlacesSelect } from '../../../packages/places-select/src';
import React from 'react';
import renderer from 'react-test-renderer';

describe('PlacesSelect', () => {
    it('should render without suggestions', () => {
        // given
        const props = {
            address: undefined,
            setAddress: jest.fn(),
        };

        // when
        const component = renderer.create(<PlacesSelect {...props} />);
        const tree = component.toJSON();

        // then
        expect(tree).toMatchSnapshot();
    });
});
