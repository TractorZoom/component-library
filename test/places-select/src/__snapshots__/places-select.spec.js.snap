// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlacesSelect should render without suggestions 1`] = `
<div
  aria-expanded={false}
  aria-owns={null}
  className="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  role="combobox"
>
  <div
    className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
      data-shrink={false}
      id="undefined-label"
    >
      Location
    </label>
    <div
      className="MuiInputBase-root MuiFilledInput-root makeStyles-input-7 MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiFilledInput-adornedEnd makeStyles-adornedEnd-2"
      onClick={[Function]}
    >
      <input
        aria-activedescendant={null}
        aria-autocomplete="both"
        aria-controls={null}
        aria-invalid={false}
        autoCapitalize="none"
        autoComplete="off"
        autoFocus={false}
        className="MuiInputBase-input MuiFilledInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused makeStyles-nativeInput-8 MuiInputBase-inputAdornedEnd MuiFilledInput-inputAdornedEnd"
        disabled={false}
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        required={false}
        spellCheck="false"
        type="text"
        value=""
      />
      <div
        className="MuiInputAdornment-root makeStyles-chevronIcon-3 MuiInputAdornment-filled MuiInputAdornment-positionEnd"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
